exporters/common/src/main/java/io/opentelemetry/exporter/internal/marshal/Serializer.java

```diff
@@ -301,6 +377,160 @@ public abstract void serializeRepeatedMessage(ProtoFieldInfo field, Marshaler[]
   public abstract void serializeRepeatedMessage(
       ProtoFieldInfo field, List<? extends Marshaler> repeatedMessage) throws IOException;
 
+  /** Serializes {@code repeated message} field. */
+  public abstract <T> void serializeRepeatedMessage(
+      ProtoFieldInfo field,
+      List<? extends T> messages,
+      StatelessMarshaler<T> marshaler,
+      MarshalerContext context)
+      throws IOException;
+
+  @SuppressWarnings("unchecked")
+  public <T> void serializeRepeatedMessage(
+      ProtoFieldInfo field,
+      Collection<? extends T> messages,
+      StatelessMarshaler<T> marshaler,
+      MarshalerContext context,
+      MarshalerContext.Key key)
+      throws IOException {
+    if (messages instanceof List) {
+      serializeRepeatedMessage(field, (List<T>) messages, marshaler, context);
+      return;
+    }
+
+    writeStartRepeated(field);
+
+    if (!messages.isEmpty()) {
+      RepeatedElementWriter<T> writer = context.getInstance(key, RepeatedElementWriter::new);
+      writer.initialize(field, this, marshaler, context);
+      messages.forEach(writer);
```

Comment:
The `RepeatedElementWriter` appears overly complicated for what seems to be accomplishable by:
```
      messages.forEach((mapKey, value) -> {
        try {
          writeStartRepeatedElement(field, context.getSize());
          marshaler.writeTo(this, mapKey, value, context);
          writeEndRepeatedElement();
        } catch (IOException e) {
          throw new IllegalStateException(e);
        }
      });
```

Reply:
As far as I understand this would be a capturing lambda that would cause an allocation.

Reply:
Huh kind of unexpected. Can punt on it for now. Maybe I'll profile the difference at some point to see if we can simplify.